[CALCITE-2450] Normalize RexCall predicates when computing digest

vlsi opened the pull request:
`$0=$1` and `$1=$0` have the same semantics, so it is worth normalizing them to reduce the planning time.

See https://issues.apache.org/jira/browse/CALCITE-2450

danny0405 commented on core/src/main/java/org/apache/calcite/rex/RexCall.java:
```diff
@@ -127,7 +143,23 @@ protected final StringBuilder appendOperands(StringBuilder sb) {
+      operandDigests.add(((RexLiteral) operand).computeDigest(includeType));
+    }
+    int totalLength = (operandDigests.size() - 1) * 2; // commas
+    for (String s : operandDigests) {
```
How about the operands are empty ?

vlsi replied:
They must not be empty, because `AND()`, `=()` and so on means nothing.

danny0405 commented:
Please **do not** merge the PR before we solve the plan compatibility.

vlsi commented:
@danny0405 , what do you mean by "solve the plan compatibility"?

danny0405 commented:
> @danny0405 , what do you mean by "solve the plan compatibility"?

The plan changes, with same semantic but no  promotion. "The plan compatibility" means we should keep as the same plan for same query/relational expression for different Calcite version.

vlsi commented:
> The plan changes, with same semantic but no promotion

The key improvement here is performance improvement.
There are several tests that get faster.
For instance: org.apache.calcite.test.RelMetadataTest#testPullUpPredicatesForExprsItr
There's an assert:

```
assertThat(inputSet.pulledUpPredicates.size(), is(18));
```

After the change it becomes `12`, so the test extracts less predicates, thus the planning becomes faster.


Note: even though the plan representation change, it is "one-time" change.
In other words, the plan would be sorted in exactly the same order, so the change would make plans **more stable** in the future.

> we should keep as the same plan for same query/relational expression for different Calcite version

It can't always be like that. A change to the optimizer might produce a slightly different plan.
This PR improves performance across different tests, so it is a true performance improvement.

However, I do not see the way to fully reproduce "old plan representation" because "old representation" is "a random one".

danny0405 commented:
How many changed plans there can have a performance promotion ? I would say very few. We can do some promotion code and that is awesome, but please **do not bring in** extra burden to the users that do not need this feature.

> In other words, the plan would be sorted in exactly the same order, so the change would make plans more stable in the future.

I don't think so, isn't keep the rex calls the sequence just as what user writes them more natural, more readable more stable ?

vlsi commented:
> isn't keep the rex calls the sequence just as what user writes them more natural, more readable more stable ?

RexSimplify can easily adjust the nodes, put elements to HashSet, and so on.

> I would say very few

How do you know that?
https://issues.apache.org/jira/browse/CALCITE-2223 is quite popular.

yanlin-Lynn commented on core/build.gradle.kts:
```diff
@@ -41,6 +41,7 @@ dependencies {
 
     api("com.fasterxml.jackson.core:jackson-annotations")
     api("org.apache.calcite.avatica:avatica-core")
+    api("org.apiguardian:apiguardian-api")
```
this seems to be another feature that is unrelated with this pr

vlsi replied:
It is related. The dependency is needed to add `@API(since = "1.22",` annotation to the methods that are added in this PR.

danny0405 replied:
Can we write our own annotation ? I would expect to not introduce a dependency just for a annotation, using our own annotations keep better compatibility and reduce conflicts.

vlsi replied:
Do you expect the same annotation to be used in Avatica and Calcite?
Does that mean the annotation should be declared in Avatica, and can be used in Calcite only after Avatica is released? It does sound too complicated.

On the other hand, https://github.com/apiguardian-team/apiguardian is a more-or-less clear dependency that has already been tested (e.g. it is used in JUnit5).

In other words, `apiguardian-api` is already available (and it can be easily used by other projects for consistent API declaration!) while Calcite's annotation is nowhere near.

danny0405 commented on core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java:
```diff
@@ -120,24 +121,23 @@ public static EnumerableHashJoin create(
+      // Lots of plans assumed the left join was hashed, so
+      // we adjust a cost accordingly
+      rowCount += leftRowCount * 1.1d;
+      return costFactory.makeCost(rowCount, 0, 0);
```
Why a hash cost is `row_cnt * 1.1` ?

vlsi replied:
The idea was hashtable creation has a linear cost, but it is a bit more than just processing the entries.

In practice, this needs adjustments, as org.apache.calcite.test.JdbcTest#testAlmostBushy plan must be flipped, but currently, it does not flip, so hash join should be adjusted somehow.

Probably I'm missing hash lookup cost.

vlsi replied:
The problem here is not 1.1, but the problem is join cardinality estimation.
Calcite tends to over-estimate the join cardinality, thus leftRowCount and rightRowCount tend to be noise when compared with the estimated join cardinality, thus all those `+= leftRowCount * 1.1d` change nothing.

So a series of patches (https://issues.apache.org/jira/browse/CALCITE-3666, https://issues.apache.org/jira/browse/CALCITE-3661) is required before this costing could provide any meaningful results :(

danny0405 commented on core/src/main/java/org/apache/calcite/config/CalciteSystemProperty.java:
```diff
@@ -83,6 +83,10 @@
       booleanProperty("calcite.enable.stream", true);
 
+  /** Whether streaming is enabled in the default planner configuration. */
+  public static final CalciteSystemProperty<Boolean> ENABLE_REX_DIGEST_NORMALIZE =
```
Fix the java doc.

vlsi replied:
fixed, thanks

danny0405 commented on core/src/main/java/org/apache/calcite/rel/externalize/RelJsonWriter.java:
```diff
@@ -140,6 +142,17 @@ public RelWriter done(RelNode node) {
+      skipNormalize = RexNode.DESCRIBE_WITHOUT_NORMALIZE.get();
+      skipNormalize.incrementAndGet();
+    }
+    try {
```
This wrapper appears in many places, can we write a method to encapsulate that ?

vlsi replied:
I would consider that to be on the hot path (especially in RexCall), so I would refrain from allocating extra objects there.

Adding a method like `withoutNormalize(Runnable)` won't make a huge readability difference, yet it would add overhead. The API is not intended for public use, so I'm inclined to leave it as is.

vlsi replied:
I've extracted that to `class RexNode { ... public static Closeable skipNormalize()` which is to be used in try-with-resources.

danny0405 commented on core/src/main/java/org/apache/calcite/rex/RexCall.java:
```diff
@@ -73,6 +85,23 @@
+    SIMPLE_COMMUTATIVE_OPS = Sets.immutableEnumSet(
+        SqlKind.AND, SqlKind.OR, SqlKind.PLUS, SqlKind.TIMES, SqlKind.EQUALS, SqlKind.NOT_EQUALS,
+        SqlKind.IS_DISTINCT_FROM, SqlKind.IS_NOT_DISTINCT_FROM,
+        SqlKind.GREATEST, SqlKind.LEAST);
```
Compared to list these `SqlKind` here, i would suggest to new a fresh new `SqlKind` named `Symmetrical` in file `SqlKind.java` which include these symmetrical operators.

vlsi replied:
Moved to `SqlKind#SYMMETRICAL` and `SqlKind#SYMMETRICAL_SAME_ARG_TYPE` (which is for cases like `PLUS`, and `TIMES` which can be re-arranged only in case the argument types are the same).
In other words, timestamp + interval vs interval + timestamp are different expressions.

danny0405 commented on core/src/main/java/org/apache/calcite/rex/RexCall.java:
```diff
@@ -155,15 +237,15 @@ private static boolean equalSansNullability(RelDataType a, RelDataType b) {
 
-  protected @Nonnull String computeDigest(boolean withType) {
+  protected @Nonnull String computeDigest(boolean withType, boolean withNormalize) {
     final StringBuilder sb = new StringBuilder(op.getName());
```
I would suggest to introduce a new config class for computing digest, something like `computeDigest(RexCall.Config)` to give the interface better extensibility.

vlsi replied:
Frankly speaking, I would refrain from extracting configuration classes and making this configurable unless there are clear use cases

vlsi replied:
I've reverted `, boolean withNormalize` argument, and replaced it with `protected static boolean needNormalize()`.

vlsi commented:
@danny0405 , I don't understand what do you mean by `needs-a-final-review` label.
The change has been tested both in isolation and with the cost changes, and it behaves well.

I've added the tests and added `try-with-resources` API which simplifies the use, and `ThreadLocal` is now `private`.

danny0405 commented:
> @danny0405 , I don't understand what do you mean by `needs-a-final-review` label.
> The change has been tested both in isolation and with the cost changes, and it behaves well.
> 
> I've added the tests and added `try-with-resources` API which simplifies the use, and `ThreadLocal` is now `private`.
> 
> So I'm inclined to merge the PR.

The PR looks good from my side, go ahead ~

danny0405 commented on core/src/main/java/org/apache/calcite/rex/RexCall.java:
```diff
@@ -127,11 +137,83 @@ protected final StringBuilder appendOperands(StringBuilder sb) {
+      }
+      // fall through: order arguments below
+    } else if (!SqlKind.SYMMETRICAL.contains(kind)
+        && (kind == kind.reverse()
```
I have left the comments in another PR https://github.com/apache/calcite/pull/1740

danny0405 commented:
Overall looks pretty good, i have left some minor comments in PR https://github.com/apache/calcite/pull/1740
